### DISCUSSION TITLE
Vue watcher equivalent in Solid

### DISCUSSION THREAD

**Discussion by targetlucked69** [solidjs/solid#846] [Q&A] [answered]
What's the equivalent of vue watchers in Solid? Is it `createEffect` or `createRenderEffect`?

Example watcher in vue

```js
const data = ref('')

watch(data, (val) => {
    // runs everytime data changes
    doSomething()
})
```

In solid, how to check if dependency has been updated?

```js
createEffect((prev) => {
    if (data() !== prev) {
        doSomething();
    }

    return data()
});
```

Am I doing it correctly? 

Also what if I want to watch multiple sources? In Vue, I can do this

```js
const data = ref('')
const data2 = ref('')
const data3 = ref('')

watch([data, data2, data3], ([val, val2, val3]) => {
    // runs everytime any of data in array changes
    doSomething()
})
```

**Comment by ryansolid** [comment, marked as answer]
Yeah.. generally don't need to be explicit on the dependencies and everything will track automatically.

But you might appreciate the `on` helper. https://www.solidjs.com/docs/latest/api#on
It is based on Vue's watch.. and works with Memos and Effects.

**Reply by targetlucked69** [reply]
Nice! This works on stores too, right?

```ts
const [state] = createStore({ someKey: { hello: 'world' } })

createEffect(on(state.someKey, () => {
    doSomething()
}, {
    defer: true
}));
```

Tried and getting error:

```ts
Uncaught TypeError: deps[i] is not a function
```

**Reply by ryansolid** [reply]
Wrap the store access in a function. Since it's the property access that is tracked.

```js
const [state] = createStore({ someKey: { hello: 'world' } })

createEffect(on(() => state.someKey, () => {
    doSomething()
}, {
    defer: true
}));
```

**Reply by targetlucked69** [reply]
Oh okay exactly like vue's watcher. Thank you very much!

**Reply by thetarnav** [reply]
One thing that sucks with `on`, is that the prettier is gonna stretch your code usually. It'll look like this:
```ts
createEffect(
  on(
    () => state.someKey,
    () => {
      doSomething();
    },
    { defer: true }
  )
);
```
